history_and_modals.md

```diff
@@ -43,13 +44,86 @@ The History API makes handling back clicks for modals rather hard problem. Among
+
+There are many precedents of a modal-like functionality on the Web Platform today:
+
+- The `dialog`'s `showModal()` mode cancels the dialog on the <kbd>Esc</kbd> key.
+- The `requestFullscreen` mode cancels the full screen on the <kbd>Esc</kbd> key
+  on desktop and on the back button on mobile.
+- The dropdown of a `input[type=date]` is canceled on the <kbd>Esc</kbd> key
+  on desktop and on the back button on mobile.
+
+These features tell us that (a) there's a clear precedent for using the back
+button for cancelling, and (b) this support is platform sensitive: it's expected
+on a mobile platform (especially Android), but not on desktop.
+
 ## Solving history and modals problem
 
 One way to solve this problem is to challenge why the History API should be
 used for modals support at all. The History API integration for a modal is,
 semantically, more similar to <kbd>Esc</kbd> handling than navigation.
 
-A possible solutions, for instance, could require a popup to be built on the
-`<dialog>` element. The browser would provide the Back button integration,
-just as it provides <kbd>Esc</kbd> support today. The dialog API would also
-need to support "save data" prompt.
+Some of the possible solutions could include:
+
+1. An extended `<dialog>` API or a new platform primitive.
+2. A back button event propagated via the <kbd>Esc</kbd> key in some cases.
+
+### An extended `<dialog>` API or a new platform primitive
+
+This solution would require a popup to be built on the `<dialog>` element. The
+browser would provide the Back button integration, just as it provides
+<kbd>Esc</kbd> support today.
+
+There are some things that dialog API would need to improve:
+1. Support more styling options to match different use cases, such as an input
+   dropdown.
+2. Support the "save data" prompt.
+
+Put it together, the API use could look like this:
+
+```javascript
+const dialog = document.createElement('dialog');
+document.body.appendChild(dialog);
+
+dialog.onclose = () => {
+  if (!allDataSaved()) {
+    return 'Are you sure you want to close?';
+  }
+}
+dialog.showModal();
+```
+
+The `onclose` handler is constructed similar to `unbeforeunload` event to ensure
+that a "save data" prompt cannot be abused as an escape hatch to block the back
+button.
+
+If Dialog API cannot be extended in this way, there could be a similar new
+HTML element that can support all of these features and allow flexible styling.
+The `dialog` element itself could extend from this new type of the element.
+
+### A back button event propagated via the Esc key
+
+The browser could automatically propagate the back button activation as an <kbd>Esc</kbd>
+key event. The main benefit of this would be that many existing popup/dialog/dropdown
+implementations that already support the <kbd>Esc</kbd> key, could automatically
+gain the back button support without any changes or only with minor updates.
+
+The use of such API could then be as simple as keyboard event handling:
```

Comment:
I think the major drawback of this API is that it's a surprise. When an event says it's an "escape key" but really it's a "back button," it just feels like that's fairly unintuitive of an API

Reply:
I tend to agree. I think a new event with separate semantics is much better. Who knows what other cancelation gestures might be available on a platform.